src/server/pkg/obj/cache_client.go

```diff
@@ -0,0 +1,149 @@
+package obj
+
+import (
+	"context"
+	"io"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/golang-lru/simplelru"
+	log "github.com/sirupsen/logrus"
+)
+
+var _ Client = &cacheClient{}
+
+type cacheClient struct {
+	slow, fast Client
+
+	mu           sync.Mutex
+	cache        *simplelru.LRU
+	populateOnce sync.Once
+}
+
+// NewCacheClient returns slow wrapped in an LRU write-through cache using fast for storing
+// cached data.
+func NewCacheClient(slow, fast Client, size int) Client {
+	if size == 0 {
+		return slow
+	}
+	client := &cacheClient{
+		slow: slow,
+		fast: fast,
+	}
+	cache, err := simplelru.NewLRU(size, client.onEvicted)
+	if err != nil {
+		// lru.NewWithEvict only errors for size < 1
+		panic(err)
+	}
+	client.cache = cache
+	return client
+}
+
+func (c *cacheClient) Reader(ctx context.Context, p string, offset, size uint64) (io.ReadCloser, error) {
+	c.doPopulateOnce(ctx)
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if _, exists := c.cache.Get(p); exists {
+		return c.fast.Reader(ctx, p, offset, size)
+	}
+	if err := Copy(ctx, c.slow, c.fast, p, p); err != nil {
+		return nil, err
+	}
+	c.cache.Add(p, struct{}{})
+	return c.fast.Reader(ctx, p, offset, size)
+}
+
+func (c *cacheClient) Writer(ctx context.Context, p string) (io.WriteCloser, error) {
+	wc, err := c.slow.Writer(ctx, p)
+	if err != nil {
+		return nil, err
+	}
+	return &cacheWriter{WriteCloser: wc, path: p, client: c}, nil
```

Comment:
I don't think the cache invalidation here is going to get us much because writes / deletes can be performed by other Pachd instances and we will not be overwriting an existing chunk. I think we should probably only use this caching client for chunks because stale reads are much less / not problematic in that narrow use case. If the other parts of our system are working correctly, then stale reads should not occur since the chunks should exist if we are trying to read them. In the case where they don't (in the slow storage), we should expect to eventually error when reading a deleted chunk in the slow storage or just succeed as if the data was not deleted yet (I think this would primarily come into play with temporary file sets).

Reply:
I agree that this is terrible general purpose object cache.  But this at least allows Read-Your-Writes behavior.  If this is living in the `obj` package I think we should do the best we can.

Reply:
I think we should just appropriately document the intended use of this caching client to make clear the limitations. I don't think cache invalidation should be supported in any fashion by this client to keep it simple. I don't think we want to get in  the territory of kind of supporting cache invalidation, but not in the general sense.